fix(solver): literal elements in a spread-containing array literal are not widened during T[] inference, masking TS2322

## Summary

When inferring the element type `T` of `arr: T[]` from an array literal that contains **both a spread and non-spread literal elements** (e.g. `[...base, "x"]`), tsz preserves the literal types of the non-spread elements (`T = number | "x"`) instead of widening them like tsc (`T = string | number`). Plain array literals (no spread) widen correctly in tsz; only the spread-plus-literal combination regresses. The wrong inferred type masks a `TS2322` (false negative).

## Minimal repro

```ts
declare function f<T>(arr: T[]): T;
const base = [1, 2];
const r = f([...base, "x"]);
const c: number | "x" = r;   // tsz: accepts (WRONG); tsc: errors
```

tsc (`--strict`):
```
error TS2322: Type 'string | number' is not assignable to type 'number | "x"'.
```

tsz: no error (it inferred `T = number | "x"`).

### Inference probe

```ts
declare function f<T>(arr: T[]): T;
const base = [1, 2];
const r = f([...base, "x"]);
const c1: string | number = r;  // tsc OK; tsz: TS2322 'number | "x"' not assignable
const c2: number | "x" = r;     // tsc: TS2322; tsz: OK
```

tsc infers `T = string | number`; tsz infers `T = number | "x"`.

## Expected

The non-spread literal elements of an array literal are widened during `T[]` inference (this is ordinary array-literal element widening), giving `T = string | number`. tsz should match tsc.

## Boundary / controls

- Plain `f(["x","y"])` (no spread): both widen to `string` — control passes.
- Plain mixed `f([1,"x"])` (no spread): both infer `string | number` — control passes.
- `f([...["a","b"]])` (spread of a literal array, no extra literal element): both widen to `string` — control passes. So the trigger is *non-spread literal elements alongside a spread*, not the spread alone.
- `f(["x", ...base])` (spread last): reproduces — position of the spread does not matter.
- `f([...base, "x", "y"])` (multiple trailing literals): reproduces.

## Root-cause hypothesis + layer

Solver array-literal element-type computation feeding inference. The literal-widening pass that applies to the elements of a fresh array literal is skipped (or the element types are taken pre-widening) when the array literal contains a spread, so spread-mixed literals retain their literal types and flow into `T`. The widening should apply to the non-spread elements regardless of whether a spread is present. Fix in `tsz-solver` array-literal type construction / element widening for spread-containing literals.

## Test matrix

1. `f([...base, "x"])` reported repro — expect `T = string | number`.
2. `f(["x", ...base])` spread last — same.
3. `f([...base, "x", "y"])` multiple literals — expect widened.
4. Control: `f(["x","y"])` plain — must keep widening (already works).
5. Control: `f([...["a","b"]])` spread-only — must keep widening.

AgentName: gen-hunter

--- a/crates/tsz-checker/src/lib.rs
+++ b/crates/tsz-checker/src/lib.rs
@@ -400,11 +400,11 @@ mod architecture_contract_tests_src;
 #[path = "../tests/array_isarray_mutual_subtype_narrowing_tests.rs"]
 mod array_isarray_mutual_subtype_narrowing_tests;
 #[cfg(test)]
-#[path = "tests/as_const_nested_literal_display_tests.rs"]
-mod as_const_nested_literal_display_tests;
-#[cfg(test)]
 #[path = "tests/array_literal_spread_inference_widening_tests.rs"]
 mod array_literal_spread_inference_widening_tests;
+#[cfg(test)]
+#[path = "tests/as_const_nested_literal_display_tests.rs"]
+mod as_const_nested_literal_display_tests;
 #[cfg(test)]
 #[path = "tests/assertion_type_predicate_diagnostics_tests.rs"]
 mod assertion_type_predicate_diagnostics_tests;

--- a/crates/tsz-checker/src/lib.rs
+++ b/crates/tsz-checker/src/lib.rs
@@ -403,6 +403,9 @@ mod array_isarray_mutual_subtype_narrowing_tests;
 #[path = "tests/as_const_nested_literal_display_tests.rs"]
 mod as_const_nested_literal_display_tests;
 #[cfg(test)]
+#[path = "tests/array_literal_spread_inference_widening_tests.rs"]
+mod array_literal_spread_inference_widening_tests;
+#[cfg(test)]
 #[path = "tests/assertion_type_predicate_diagnostics_tests.rs"]
 mod assertion_type_predicate_diagnostics_tests;
 #[cfg(test)]

--- a/crates/tsz-checker/src/tests/array_literal_spread_inference_widening_tests.rs
+++ b/crates/tsz-checker/src/tests/array_literal_spread_inference_widening_tests.rs
@@ -1,0 +1,159 @@
+//! Tests for issue #9741: fresh literal elements of a spread-containing array
+//! literal must be widened during `T[]` inference.
+//!
+//! When inferring the element type `T` of `arr: T[]` from an array literal that
+//! mixes a spread with non-spread literal elements (e.g. `[...base, "x"]`), the
+//! inferred element-type candidate is a single union (`number | "x"`). tsc's
+//! `getWidenedLiteralType` widens the fresh literal members of that union
+//! independently, yielding `T = string | number`. The bug preserved the literal
+//! members (`T = number | "x"`), masking a TS2322.
+//!
+//! The structural rule: when a type parameter is inferred from array-literal
+//! element types and the inferred element type is a union containing fresh
+//! literal members, those fresh members are widened even when the union also
+//! carries non-literal members.
+
+use crate::test_utils::check_source_diagnostics;
+
+fn ts2322(src: &str) -> Vec<String> {
+    check_source_diagnostics(src)
+        .into_iter()
+        .filter(|d| d.code == 2322)
+        .map(|d| d.message_text)
+        .collect()
+}
+
+#[test]
+fn spread_then_literal_widens_during_array_inference() {
+    // `T` widens to `string | number`, so the narrow literal target is rejected.
+    let errs = ts2322(
+        r#"
+declare function f<T>(arr: T[]): T;
+const base = [1, 2];
+const r = f([...base, "x"]);
+const c: number | "x" = r;
+"#,
+    );
+    assert_eq!(errs.len(), 1, "expected TS2322, got: {errs:?}");
+}
+
+#[test]
+fn spread_then_literal_accepts_widened_target() {
+    // The widened element type `string | number` must be accepted.
+    let errs = ts2322(
+        r#"
+declare function f<T>(arr: T[]): T;
+const base = [1, 2];
+const r = f([...base, "x"]);
+const c: string | number = r;
+"#,
+    );
+    assert!(errs.is_empty(), "expected no TS2322, got: {errs:?}");
+}
+
+#[test]
+fn literal_then_spread_widens_regardless_of_position() {
+    let errs = ts2322(
+        r#"
+declare function pick<U>(items: U[]): U;
+const nums = [1, 2];
+const r = pick(["x", ...nums]);
+const c: number | "x" = r;
+"#,
+    );
+    assert_eq!(errs.len(), 1, "expected TS2322, got: {errs:?}");
+}
+
+#[test]
+fn spread_with_multiple_trailing_literals_widens() {
+    let errs = ts2322(
+        r#"
+declare function f<T>(arr: T[]): T;
+const base = [1, 2];
+const r = f([...base, "x", "y"]);
+const c: number | "x" = r;
+"#,
+    );
+    assert_eq!(errs.len(), 1, "expected TS2322, got: {errs:?}");
+}
+
+#[test]
+fn spread_with_boolean_literal_widens_to_boolean() {
+    // Adjacent shape: a boolean literal mixed with a numeric spread widens to
+    // `number | boolean`, not `number | true`.
+    let errs = ts2322(
+        r#"
+declare function f<T>(arr: T[]): T;
+const base = [1, 2];
+const r = f([...base, true]);
+const c: number | true = r;
+"#,
+    );
+    assert_eq!(errs.len(), 1, "expected TS2322, got: {errs:?}");
+}
+
+#[test]
+fn plain_literal_array_still_widens_control() {
+    // Control: a plain literal array (no spread) must keep widening to `string`.
+    let errs = ts2322(
+        r#"
+declare function f<T>(arr: T[]): T;
+const r = f(["x", "y"]);
+const c: "x" | "y" = r;
+"#,
+    );
+    assert_eq!(errs.len(), 1, "expected TS2322, got: {errs:?}");
+}
+
+#[test]
+fn spread_of_literal_array_still_widens_control() {
+    // Control: spreading a literal-typed array (its elements already widened to
+    // `string`) plus another literal still widens.
+    let errs = ts2322(
+        r#"
+declare function f<T>(arr: T[]): T;
+const sbase = ["a", "b"];
+const r = f([...sbase, "c"]);
+const c: "a" | "b" | "c" = r;
+"#,
+    );
+    assert_eq!(errs.len(), 1, "expected TS2322, got: {errs:?}");
+}
+
+#[test]
+fn const_type_param_preserves_literals_negative_control() {
+    // Negative control: a `const` type parameter must still preserve literal
+    // element types (no widening), so the narrow target is accepted.
+    let errs = ts2322(
+        r#"
+declare function f<const T>(arr: T[]): T;
+const base = [1, 2] as const;
+const r = f([...base, "x"]);
+const c: 1 | 2 | "x" = r;
+"#,
+    );
+    assert!(
+        errs.is_empty(),
+        "expected no TS2322 for const T, got: {errs:?}"
+    );
+}
+
+#[test]
+fn nullable_literal_array_keeps_literals_boundary() {
+    // Boundary: a plain literal array whose element union also carries a
+    // nullable (`null`) member is left un-widened. The fix only widens when the
+    // non-literal members are widened primitives (the spread-of-widened-array
+    // shape); nullable-mixed unions keep their literal members, so the literal
+    // target is still accepted.
+    let errs = ts2322(
+        r#"
+declare function f<T>(arr: T[]): T;
+const r = f([true, 1, null, "yes"]);
+const c: true | 1 | "yes" | null = r;
+"#,
+    );
+    assert!(
+        errs.is_empty(),
+        "expected no TS2322 for nullable-mixed literal array, got: {errs:?}"
+    );
+}

--- a/crates/tsz-solver/src/inference/infer.rs
+++ b/crates/tsz-solver/src/inference/infer.rs
@@ -14,7 +14,7 @@ use crate::construction::TypeDatabase;
 #[cfg(test)]
 use crate::types::*;
 use crate::types::{InferencePriority, TemplateSpan, TypeData, TypeId};
-use crate::visitor::{is_literal_type, is_union_of_fresh_literals};
+use crate::visitor::{array_element_union_widens_literals, is_literal_type};
 use ena::unify::{InPlaceUnificationTable, NoError, UnifyKey, UnifyValue};
 use rustc_hash::{FxHashMap, FxHashSet};
 use std::cell::RefCell;
@@ -1252,7 +1252,7 @@ impl<'a> InferenceContext<'a> {
             is_fresh_literal: (!context.from_object_property || context.source_is_fresh)
                 && (is_literal_type(self.interner, ty)
                     || (self.in_array_element_context
-                        && is_union_of_fresh_literals(self.interner, ty)))
+                        && array_element_union_widens_literals(self.interner, ty)))
                 && !self.source_is_type_annotation
                 && !self.in_readonly_source_context,
             from_object_property: context.from_object_property,

--- a/crates/tsz-solver/src/visitors/visitor_predicates.rs
+++ b/crates/tsz-solver/src/visitors/visitor_predicates.rs
@@ -92,6 +92,48 @@ pub fn is_union_of_fresh_literals(types: &dyn TypeDatabase, type_id: TypeId) -> 
     }
 }
 
+/// Decide whether an array-element inference union should have its fresh literal
+/// members widened to their primitive base.
+///
+/// Returns `true` when the union has at least one literal member and every
+/// member is either a literal or one of the primitives that literal widening
+/// produces (`number` / `string` / `boolean` / `bigint`). This covers:
+/// - pure literal unions (`"a" | "b"`, `1 | 2`) — equivalent to
+///   `is_union_of_fresh_literals`; and
+/// - unions that mix fresh literals with an already-widened primitive, which is
+///   exactly the shape produced by spreading a widened array alongside a literal
+///   element (`number | "x"` from `[...numberArray, "x"]`). The widened
+///   primitive proves the array literal already carries a widened element, so
+///   the fresh literal siblings must widen too, matching tsc's
+///   `getWidenedLiteralType`.
+///
+/// A union whose non-literal members include `null` / `undefined` / objects is
+/// left alone, preserving the literal members for downstream narrowing (the
+/// conservative baseline for mixed literal+nullable element unions).
+pub fn array_element_union_widens_literals(types: &dyn TypeDatabase, type_id: TypeId) -> bool {
+    if type_id.is_intrinsic() {
+        return false;
+    }
+    match types.lookup(type_id) {
+        Some(TypeData::Union(list_id)) => {
+            let members = types.type_list(list_id);
+            let mut has_literal = false;
+            for &member in members.iter() {
+                if is_literal_type(types, member) {
+                    has_literal = true;
+                } else if !matches!(
+                    member,
+                    TypeId::NUMBER | TypeId::STRING | TypeId::BOOLEAN | TypeId::BIGINT
+                ) {
+                    return false;
+                }
+            }
+            has_literal
+        }
+        _ => false,
+    }
+}
+
 /// Check if a type is a module namespace type (import * as ns).
 ///
 /// Matches: `TypeData::ModuleNamespace`(_)

